src/compiler/checker.ts

```diff
@@ -383,20 +383,42 @@ namespace ts {
             // return undefined if we can't find a symbol.
         }
 
-        /** Returns true if node1 is defined before node 2**/
-        function isDefinedBefore(node1: Node, node2: Node): boolean {
-            let file1 = getSourceFileOfNode(node1);
-            let file2 = getSourceFileOfNode(node2);
+        const enum RelativeLocation {
+            Unknown,
+            SameFileLocatedBefore,
+            SameFileLocatedAfter,
+            DifferentFilesLocatedBefore,
+            DifferentFilesLocatedAfter,
+        }
+
+        function isLocatedBefore(origin: Node, target: Node): boolean {
+            switch (getRelativeLocation(origin, target)) {
+                // unknown is returned with nodes are in different files and order cannot be determined based on compilation settings
+                // optimistically assume this is ok
+                case RelativeLocation.Unknown:
+                case RelativeLocation.SameFileLocatedBefore:
+                case RelativeLocation.DifferentFilesLocatedBefore:
+                    return true;
+                default:
+                    return false;
+            }
+        }
+
+        /** gets relative location of target comparing to origin **/
+        function getRelativeLocation(origin: Node, target: Node): RelativeLocation {
+            let file1 = getSourceFileOfNode(origin);
+            let file2 = getSourceFileOfNode(target);
             if (file1 === file2) {
-                return node1.pos <= node2.pos;
+                return origin.pos > target.pos ? RelativeLocation.SameFileLocatedBefore : RelativeLocation.SameFileLocatedAfter;
             }
 
-            if (!compilerOptions.outFile && !compilerOptions.out) {
-                return true;
+            if (modulekind || (!compilerOptions.outFile && !compilerOptions.out)) {

```

Comment:
What if moduleKind is specified, but -out is also specified, and both files are part of the global script? Should it still be unknown?